### PR TITLE
GLFW allow passing multiple icons

### PR DESCRIPTION
**Pull request by musm**

### CODE REVIEW

File: src/glfw3.jl [hunk 1]
```diff
@@ -509,9 +516,12 @@ GLFW.SetWindowIcon(win, buff)
 GLFW.PollEvents() # seems to need a poll events to become active
 ```
 """
-SetWindowIcon(window::Window, image::Matrix{NTuple{4, UInt8}}) = ccall((:glfwSetWindowIcon, lib),
+SetWindowIcon(window::Window, image::AbstractMatrix{NTuple{4,UInt8}}) = ccall((:glfwSetWindowIcon, lib),
```
Comment (jayschwa) [inline src/glfw3.jl]: I suggest getting rid of the `ccall` and just calling the new plural variant:
```
SetWindowIcon(window, image) = SetWindowIcon(window, [image])
```

Comment (musm) [reply src/glfw3.jl]: `ERROR: LoadError: GLFWError (PLATFORM_ERROR): Win32: Failed to create RGBA bitmap`

when trying your suggestions and passing only when icon

I will drop the last commit

Comment (musm) [reply src/glfw3.jl]: resolved

**Comment by musm** [comment]
@jayschwa   I am very happy to have finally fixed the subtle bug.  I think this is now in shape to be merged.

**Comment by musm** [comment]
The logic is also now *vastly* simpler and probably safer than previous

File: src/glfw3.jl [hunk 2]
```diff
@@ -389,17 +389,11 @@ Base.show(io::IO, vm::VidMode) = write(io, "$(vm.width)x$(vm.height)@$(vm.refres
...
-function Base.cconvert(::Type{GLFWImage}, image::Matrix{NTuple{4, UInt8}})
-	ptr = Base.cconvert(Ptr{UInt8}, image)
-	(size(image)..., ptr)
-end
-
-function Base.unsafe_convert(::Type{GLFWImage}, data::Tuple{Int, Int, Matrix{NTuple{4, UInt8}}})
-	ptr = Base.unsafe_convert(Ptr{UInt8}, data[3])
-	GLFWImage(Cint(data[1]), Cint(data[2]), ptr)
+function Base.convert(::Type{GLFWImage}, image::AbstractMatrix{NTuple{4,UInt8}})
+	GLFWImage(size(image)...,  Base.unsafe_convert(Ptr{UInt8}, image))
```
Comment (musm) [inline src/glfw3.jl]: I think there may be an exceptional case that the GC can erase `image`,  and thus this reference would then be illegal.

**Comment by musm** [comment]
@jayschwa  thanks, let me investigate further to see if I can fix the unsafe behavior.

File: src/glfw3.jl (hunk 2 above)
Comment (SimonDanisch) [reply src/glfw3.jl]: yeah that looks wrong... I'm pretty sure you need to cascade cconvert & unsafe_convert in some way, to make ccall root image! maybe not like i did, but i think you should at least call unsafe_convert only on arguments to cconvert!

File: src/glfw3.jl [hunk 3]
```diff
@@ -509,8 +510,11 @@ GLFW.SetWindowIcon(win, buff)
 GLFW.PollEvents() # seems to need a poll events to become active
 ```
 """
-SetWindowIcon(window::Window, image::Matrix{NTuple{4, UInt8}}) = ccall((:glfwSetWindowIcon, lib),
-	Cvoid, (Window, Cint, GLFWImage), window, 1, image)
```
Comment (musm) [inline src/glfw3.jl]: I'm really surprised this code even worked ...    the c signature requires a pointer to GLFWImage, however the julia code here passes directly a GLFWImage struct to  the c function.

Comment (SimonDanisch) [reply src/glfw3.jl]: cconvert should do the magic ;)

Comment (musm) [reply src/glfw3.jl]: right, but the manual overrides ensured that  a   GLFW struct is directly passes to the signature

```
Base.unsafe_convert(GLFWImage, Base.cconvert(GLFWImage, image))
```
return a GLFWImage which is then passed directly to glfwSetWindowIcon, but this function takes as inputs  pointer to an array with the GLFWImage structure.....

**Comment by musm** [comment]
I think I am done here,  please see the last commit which is  pretty optional, and I wouldn't have a problem dropping it.

File: src/glfw3.jl (hunk 3 above)
Comment (SimonDanisch) [reply src/glfw3.jl]: Hm I see... maybe should have been wrapped in a Ref, and we got lucky that structs get passed by reference for ccall in any case ;)

Comment (musm) [reply src/glfw3.jl]: `Hm I see... maybe should have been wrapped in a Ref, and we got lucky that structs get passed by reference for ccall in any case ;)

I don't think wrapping in a `Ref` would work, but now that you mention that structs get passed by reference,  I suppose this is why it worked....

Comment (musm) [reply src/glfw3.jl]: >  structs get passed by reference for ccall 

is this documented?  I can't find it here https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/  but I belive you

**Comment by jayschwa** [comment]
I just tried the example code in the docstring, and it doesn't work for me in Julia 1.0:
```julia
julia> using FileIO

julia> icons = load(["red_gem.png", "chest.png"])
ERROR: MethodError: no method matching load(::Array{String,1})
Closest candidates are:
  load(::Union{AbstractString, IO}, ::Any...; options...) at C:\Users\jaysc\.julia\packages\FileIO\lt9I0\src\loadsave.jl:118
  load(::Formatted{F}, ::Any...; options...) where F at C:\Users\jaysc\.julia\packages\FileIO\lt9I0\src\loadsave.jl:172
```

**Comment by musm** [comment]
Should be `load.(.....`

Hold of from tagging just yet

**Pull request merged**